### DISCUSSION TITLE
Model - coerce string to dict for field

### DISCUSSION THREAD

**Discussion by WilliamStam** [pydantic/pydantic#8446] [Question] [answered]
If im retrieving data as a dict but there are some fields in the dict that are nested dicts. if a field is set to be type dict and the data coming in is a str for that field what is the best way to coerce it to be a dict?

```
import pydantic

input = {
    "id": 4,
    "data": '{"something":4,"field":"else"}'
}
class MyModel(pydantic.BaseModel):
    id: int
    data: dict = pydantic.Field(default_factory=dict)

model = MyModel.model_validate(input)
print(model)
```

results in a 

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for MyModel
data
Input should be a valid dictionary [type=dict_type, input_value='{"something":4,"field":"else"}', input_type=str]
```

i get why its doing it. and its perfectly right to be doing so. what would be the best way to check that the field is set to be a dict (

i did try
```
j = json.dumps(input)
model = MyModel.model_validate_json(j)
```

but this turns the json into

```
{"id": 4, "data": "{\"something\":4,\"field\":\"else\"}"}
```

the "easy" fix is to convert the field to json via json.loads() before passing it to the model but then i have to keep both places (model and maker of the model) in check. 

ie:
```
input['data'] = json.loads(input['data'])
model = MyModel.model_validate(input)
print(model)
```

i have some pretty big models so would prefer not doing this if at all possible or at the very least do something like `for field in MyModel: if field type is dict and input is str then json.loads()`

suggestions?

**Comment by Kludex** [comment, marked as answer]
Is this enough?

```py
import pydantic
from typing import Any
from pydantic.types import Json

input = {
    "id": 4,
    "data": '{"something":4,"field":"else"}'
}
class MyModel(pydantic.BaseModel):
    id: int
    data: Json[Any] = pydantic.Field(default_factory=dict)  # <<<

model = MyModel.model_validate(input)
print(model)
```

**Reply by WilliamStam** [reply]
wow thank you!!!

```
data: Json[Any] | None = pydantic.Field(default_factory=dict)
```
works great!
